[bug] Dataset preparation fails

# 🐞 Describe the Bug

Dataset preparation fails with the following error, for some datasets it succeeds if `wrold_size=1`
```
Saving shard 0: 100%|██████████| 1169149/1169149 [01:24<00:00, 13904.80docs/s]
2025-03-06 17:47:47,543 Traceback (most recent call last):
  File "/app/fast_llm/tools/cli.py", line 29, in fast_llm
    Runnable.parse_and_run(unparsed)
  File "/app/fast_llm/engine/config_utils/runnable.py", line 36, in parse_and_run
    runnable()
  File "/app/fast_llm/data/preparator/gpt_memmap/prepare.py", line 265, in run
    self._blend_dataset_configs(dataset_configs), self._config.output_path / f"fast_llm_config.yaml"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/fast_llm/data/preparator/gpt_memmap/prepare.py", line 288, in _blend_dataset_configs
    return GPTIndexedDatasetConfig.from_dict(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/fast_llm/config.py", line 683, in from_dict
    return cls._from_dict(default, strict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/fast_llm/data/dataset/gpt/config.py", line 113, in _from_dict
    Assert.custom(issubclass, actual_cls, cls)
  File "/app/fast_llm/utils.py", line 178, in custom
    assert fn(
           ^^^
AssertionError: Assertion failed: fn(<class 'fast_llm.data.dataset.gpt.config.GPTBlendedDatasetConfig'>, <class 'fast_llm.data.dataset.gpt.config.GPTIndexedDatasetConfig'>)
```

# 🔄 Steps to Reproduce

Steps to reproduce the behaviour:

1. Use `sha-93d7e7d` version of Fast-LLM 
2. **Run the following command** using internal `fml-ops`:

    ``` 
        fml --multirun hydra/launcher=toolkit \
          +fast_llm_prepare=base \
          ++cmd.config.dataset.field=text \
          ++cmd.config.dataset.split=train \
          ++cmd.config.dataset.path=stas/openwebtext-10k  \
          ++cmd.config.output_path=/mnt/datasets/tokenized/Mistral-7B-v0.1-Tokenizer-Files/fast_llm_test \
          ++cmd.config.tokenizer.path=/mnt/datasets/tokenizer/Mistral-7B-v0.1-Tokenizer-Files/ \
          ++cmd.config.dataset.trust_remote_code=true \
          ++job.experiment_name=fast-llm-prepare-test \
          ++job.base_image.version_tag=sha-93d7e7d \
          +data_ids=yul_201_fml
        
    ```

# 🎯 Expected Behavior
Succesfull dataset preparation.

--- a/fast_llm/data/preparator/gpt_memmap/prepare.py
+++ b/fast_llm/data/preparator/gpt_memmap/prepare.py
@@ -19,6 +19,7 @@ from fast_llm.data.dataset.gpt.config import (
     GPTDatasetSliceConfig,
     GPTIndexedDatasetConfig,
     GPTMemmapDatasetConfig,
+    GPTSampledDatasetConfig,
 )
 from fast_llm.data.dataset.gpt.memmap import GPTMemmapDataset
 from fast_llm.data.dataset.gpt.sampled import GPTSample
@@ -255,7 +256,7 @@ class GPTMemmapDatasetPreparator[ConfigType: GPTMemmapDatasetPreparatorConfig](D
             # Create the config file(s) on rank 0
             if self._config.splits:
                 for split_name, split_config in self._split_and_blend_dataset_configs(
-                    dataset_configs, self._config.splits
+                    dataset_configs, self._config.splits, self._config.output_path
                 ).items():
                     self._save_dataset_config(
                         split_config, self._config.output_path / f"fast_llm_config_{split_name}.yaml"
@@ -285,7 +286,7 @@ class GPTMemmapDatasetPreparator[ConfigType: GPTMemmapDatasetPreparatorConfig](D
     def _blend_dataset_configs(cls, dataset_configs: list[GPTMemmapDatasetConfig]) -> GPTIndexedDatasetConfig:
         if len(dataset_configs) == 1:
             return dataset_configs[0]
-        return GPTIndexedDatasetConfig.from_dict(
+        return GPTSampledDatasetConfig.from_dict(
             {
                 "type": "blended",
                 "datasets": dataset_configs,
@@ -295,8 +296,8 @@ class GPTMemmapDatasetPreparator[ConfigType: GPTMemmapDatasetPreparatorConfig](D
 
     @classmethod
     def _split_and_blend_dataset_configs(
-        cls, dataset_configs: list[GPTMemmapDatasetConfig], splits: dict[str, int | float]
-    ) -> dict[str, GPTIndexedDatasetConfig]:
+        cls, dataset_configs: list[GPTMemmapDatasetConfig], splits: dict[str, int | float], output_path: pathlib.Path
+    ) -> dict[str, GPTSampledDatasetConfig]:
         split_cumsum = padded_cumsum(normalize_probabilities(list(splits.values()), return_array=True)).tolist()
         dataset_sizes = [dataset_config.num_tokens for dataset_config in dataset_configs]
         dataset_probabilities = normalize_probabilities(dataset_sizes)
@@ -323,7 +324,9 @@ class GPTMemmapDatasetPreparator[ConfigType: GPTMemmapDatasetPreparatorConfig](D
                     dataset_tokens_in_split.append(dataset_sizes[dataset_index])
                 elif split_end_in_dataset > split_begin_in_dataset:
                     # Part of the dataset belongs to the split.
-                    sizes_cumsum = dataset_config.build().get_document_sizes().cumsum()
+                    # TODO: Somehow getting a segfault when merging two lines below (numpy bug?).
+                    dataset = dataset_config.to_copy({"path": output_path / dataset_config.path}).build()
+                    sizes_cumsum = dataset.get_document_sizes().cumsum()
                     Assert.eq(sizes_cumsum[-1], dataset_config.num_tokens)
                     begin_index = _get_nearest_split(sizes_cumsum, split_begin_in_dataset * dataset_config.num_tokens)
                     end_index = _get_nearest_split(sizes_cumsum, split_end_in_dataset * dataset_config.num_tokens)
@@ -366,4 +369,4 @@ def _get_nearest_split(cumsum: np.ndarray, value: float) -> int:
     left = cumsum.searchsorted(value, side="right")
     if left == len(cumsum):
         return left.item()
-    return left + 1 if (value - cumsum[left]) / (cumsum[left + 1] - cumsum[left]) > 0.5 else left
+    return left.item() + 1 if (value - cumsum[left]) / (cumsum[left + 1] - cumsum[left]) > 0.5 else left.item()

--- a/tests/data/common.py
+++ b/tests/data/common.py
@@ -183,6 +183,7 @@ class MockGPTMemmapDatasetConfig(GPTIndexedDatasetConfig):
         desc="Expected number of tokens in the dataset.",
         hint=FieldHint.optional,
     )
+    path: pathlib.Path = Field(default=".")
 
     def build(self) -> "GPTIndexedDataset":
         return MockGPTMemmapDataset(self)

--- a/tests/data/test_prepare_gpt_memmap.py
+++ b/tests/data/test_prepare_gpt_memmap.py
@@ -11,6 +11,7 @@ from fast_llm.data.dataset.gpt.sampled import GPTSample
 from fast_llm.data.preparator.gpt_memmap.config import MEMMAP_DTYPES, GPTMemmapDatasetPreparatorConfig
 from fast_llm.data.preparator.gpt_memmap.prepare import GPTMemmapDatasetPreparator
 from fast_llm.utils import Assert
+from tests.data.common import MockGPTMemmapDatasetConfig  # Noqa
 
 
 def get_preparator(output_path: str, dataset_path_name: str) -> GPTMemmapDatasetPreparator:
@@ -92,6 +93,7 @@ def test_split_dataset():
     config = GPTMemmapDatasetPreparator._split_and_blend_dataset_configs(
         [dataset_config_0],
         {"training": 3, "validation": 1},
+        pathlib.Path("."),
     )
     config = {key: value.to_serialized() for key, value in config.items()}
 
@@ -120,6 +122,7 @@ def test_split_datasets_0():
     config = GPTMemmapDatasetPreparator._split_and_blend_dataset_configs(
         [dataset_config_0, dataset_config_1],
         {"training": 1, "validation": 1},
+        pathlib.Path("."),
     )
     config = {key: value.to_serialized() for key, value in config.items()}
 
@@ -136,8 +139,7 @@ def test_split_datasets_1():
     dataset_config_0 = GPTIndexedDatasetConfig.from_dict(DATASET_DICT_0.copy())
     dataset_config_1 = GPTIndexedDatasetConfig.from_dict(DATASET_DICT_1.copy())
     config = GPTMemmapDatasetPreparator._split_and_blend_dataset_configs(
-        [dataset_config_0, dataset_config_1],
-        {"training": 3, "validation": 1},
+        [dataset_config_0, dataset_config_1], {"training": 3, "validation": 1}, pathlib.Path(".")
     )
     config = {key: value.to_serialized() for key, value in config.items()}
 

